File: src_client/platform_ios/iosPluginData.m
#import "fuhaha-Swift.h"

#include "native.h"
#include "gamePluginUtil.h"
#include "gamePluginData.h"

// ----------------------------------------------------------------
// ----------------------------------------------------------------
// ----------------------------------------------------------------

// HTTP通信 コールバックバッファは要解放
void platformPluginDataHttp(void *param, char *url, char *request, void(*callback)(void *param, void *buff, size_t size)){
	int callbackId = gamePluginUtilCallbackSet(param, callback);
	NSString *nsstr1 = [NSString stringWithCString: url encoding: NSUTF8StringEncoding];
	NSString *nsstr2 = [NSString stringWithCString: request encoding: NSUTF8StringEncoding];
	[IosPluginData platformPluginDataHttp: callbackId url: nsstr1 request: nsstr2];
}

// ----------------------------------------------------------------

// ローカルデータ読み込み コールバックバッファは要解放
void platformPluginDataLocal(void *param, char *src, void(*callback)(void *param, void *buff, size_t size)){
	int callbackId = gamePluginUtilCallbackSet(param, callback);
	NSString *nsstr1 = [NSString stringWithCString: src encoding: NSUTF8StringEncoding];
	[IosPluginData platformPluginDataLocal: callbackId src: nsstr1];
}

// ----------------------------------------------------------------

// 設定読み込み 返値文字列は揮発性バッファで解放禁止
char *platformPluginDataPreferenceGet(char *key){
	NSString *nsstr1 = [NSString stringWithCString: key encoding: NSUTF8StringEncoding];
	NSString *nsstr2 = [IosPluginData platformPluginDataPreferenceGet: nsstr1];

	char* value = (char*)[nsstr2 UTF8String];
	if(value == NULL){return NULL;}
	char *buff = (char*)corePluginUtilTemporaryBuffer(strlen(value) + 1);
	strcpy(buff, value);

	return buff;
}

// 設定書き込み
void platformPluginDataPreferenceSet(char *key, char *value){
	NSString *nsstr1 = [NSString stringWithCString: key encoding: NSUTF8StringEncoding];
	NSString *nsstr2 = [NSString stringWithCString: value encoding: NSUTF8StringEncoding];
	[IosPluginData platformPluginDataPreferenceSet: nsstr1 value: nsstr2];
}

// 設定保存
void platformPluginDataPreferenceCommit(void){
	[IosPluginData platformPluginDataPreferenceCommit];
}

// ----------------------------------------------------------------
// ----------------------------------------------------------------
// ----------------------------------------------------------------

File: src_client/platform_ios/IosPluginData.swift
import Foundation

// ----------------------------------------------------------------
// ----------------------------------------------------------------
// ----------------------------------------------------------------

// プラグインクラス
class IosPluginData: NSObject{
	// ----------------------------------------------------------------

	// HTTP通信 
	static internal func platformPluginDataHttp(callbackId: Int32, url: String, request: String){
		IosPluginUtil.nativePluginUtilLoadingIncrement();

		// 通信準備
		let baseUrl: String = String.fromCString(UnsafePointer<CChar>(gamePluginUtilUrlGet()))!;
		let nsurl: NSURL = NSURL(string: baseUrl + "/" + url)!;
		let nsreq: NSMutableURLRequest = NSMutableURLRequest(URL: nsurl);
		nsreq.HTTPMethod = "POST";
		nsreq.HTTPBody = request.dataUsingEncoding(NSUTF8StringEncoding);
		nsreq.addValue("application/octet-stream", forHTTPHeaderField: "Content-Type");
		let task = NSURLSession.sharedSession().dataTaskWithRequest(nsreq, completionHandler: {(nsdat: NSData?, nsres: NSURLResponse?, error: NSError?) in
			var resData: NSData? = nsdat;
			// エラー確認
			let statusCode: Int? = (nsres as? NSHTTPURLResponse)?.statusCode;
			if(statusCode != 200){resData = nil;}
			//if(resData == nil){resData = "failed".dataUsingEncoding(NSUTF8StringEncoding);}
	
			// データ処理
			var size: Int = 0;
			var buff: UnsafeMutablePointer<CChar> = nil;
			if(resData != nil){
				size = resData!.length;
				buff = UnsafeMutablePointer<CChar>.alloc(size)
				resData!.getBytes(buff, length: size);
			}
			dispatch_async(dispatch_get_main_queue(), {
				IosPluginUtil.nativePluginUtilLoadingDecrement();
				gamePluginDataCallbackCall(callbackId, buff, size);
			});
		});
		task.resume();
	}

	// ローカルデータ読み込み
	static internal func platformPluginDataLocal(callbackId: Int32, src: String){
		IosPluginUtil.nativePluginUtilLoadingIncrement();

		dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), {
			let path: String = NSBundle.mainBundle().pathForResource(src, ofType: nil, inDirectory: "assets")!;
			let resData: NSData? = NSData(contentsOfFile: path);
			//if(resData == nil){resData = "failed".dataUsingEncoding(NSUTF8StringEncoding);}

			// データ処理
			var size: Int = 0;
			var buff: UnsafeMutablePointer<CChar> = nil;
			if(resData != nil){
				size = resData!.length;
				buff = UnsafeMutablePointer<CChar>.alloc(size)
				resData!.getBytes(buff, length: size);
			}
			dispatch_async(dispatch_get_main_queue(), {
				IosPluginUtil.nativePluginUtilLoadingDecrement();
				gamePluginDataCallbackCall(callbackId, buff, size);
			});
		});
	}

	// ----------------------------------------------------------------
}

// ----------------------------------------------------------------
// ----------------------------------------------------------------
// ----------------------------------------------------------------

